### DISCUSSION TITLE
Can I keep image loaded for secondary commands after first?

### DISCUSSION THREAD

**Discussion by jmaeding** [ImageMagick/ImageMagick#6754] [Help] [answered]
### ImageMagick version

7.0.5-Q8

### Operating system, version and so on

windows 10

### Description

I've used IM for many things, but always with one call that loaded the image and did its thing.
I now have a program that needs to get the RGB of a pixel at some coordinate, then many others but not all at once.
The images I am dealing with are large (60kx60k pixels), and I might be sampling 10 million pixels from them.
I use those to color points in a point cloud for engineering purposes.
So, is there a way to have IM load an image, then I feed it params in some loop, get the return, then more params, get return...and so on?
The reason I am trying to use IM is its so robust and memory efficient.
I am programming in C# so other approaches may be needed, like stream the pixels and just grab color values from ones I need along the way, so memory does not get full. Its just any tool must load the image at some point and you only want to do that once on these large ones.

**Comment by snibgo** [comment, marked as answer]
When a command finishes, it releases anything it has in memory.

A command can write an image to a `.mpc` file. Reading the image from that file will be fast -- it is memory-mapped, or something. An alternative is a `.miff` format, which I have found (on Windows) can be faster than mpc.

Of course, a program that you write can do anything you want, while it is running. Why do you want to end it, and re-start it, reading the same image?

**Reply by jmaeding** [reply]
wow, fast reply, thanks!
I don't want to stop and restart my program, it would be all one run.
I just needed to do many "get pixel color" calls in a loop, rather than make some huge command line call for a million points.
It really could not be put in one call anyway.
I found this on stack overflow:
[https://stackoverflow.com/questions/6020406/travel-through-pixels-in-bmp/6094092#6094092](url)
and tested on a large image. Still took a while but its within the realm of decent performance.
Shoot, just tested that on my largest image and got an overflow error.
That is the thing about IM, it chews on huge images for breakfast.

**Comment by snibgo** [comment]
> I don't want to stop and restart my program, it would be all one run.

Then that would be one command, not a "secondary command after the first".

**Reply by jmaeding** [reply]
I guess I was calling them commands because I think of IM as a command line tool. Each "convert" action would be a command.
All this command stuff is maybe too oppressive. How about "courteous request" instead?
We built this courteous request line interface tool....
Its  a lot to say, but maybe then the computer would actually do what I'm asking it, ha ha.

**Comment by fmw42** [comment]
In Imagemagick you can combine commands into one longer command and save image at any stage using parenthesis processing (and +write image.suffix) or by chaining command (with only one convert or magick call).  See https://imagemagick.org/Usage/basics/#parenthesis

**Reply by jmaeding** [reply]
In my case, the command would be like a long boring novel if it worked.
I'm not modifying an image, just extracting RGB values, millions of them.
I've done complex chaining like that but other approaches are needed.
My use of bitmap object in .net is failing as it has a max pixel width I am exceeding.
thx though!

**Comment by snibgo** [comment]
> How about "courteous request" instead?

Ha!

A "command" is something we issue at the CLI, Command Line Interface. See https://en.wikipedia.org/wiki/Command-line_interface . Within a "magick" command, we have any number of "options".
